dll/win32/msgina/msgina.c

```diff
@@ -161,6 +161,55 @@ ChooseGinaUI(VOID)
     HeapFree(GetProcessHeap(), 0, SystemStartOptions);
 }
 
+static BOOL
+SafeGetUnicodeString(const LSA_UNICODE_STRING *pInput, PWSTR pszOutput, SIZE_T cchMax)
+{
+    if (pInput && pszOutput)
+    {
+        PWSTR pszInput = pInput->Buffer;
+        SIZE_T cchInput = pInput->Length / sizeof(WCHAR);
+        if (cchInput < cchMax && (pszInput || !cchInput))
+        {
+            CopyMemory(pszOutput, pszInput, cchInput * sizeof(WCHAR));
+            pszOutput[cchInput] = UNICODE_NULL;
+            return TRUE;
+        }
+    }
+    if (pszOutput && cchMax)
+        *pszOutput = UNICODE_NULL;
+    return FALSE;
+}
+
+static BOOL
+GetLsaDefaultPassword(PGINA_CONTEXT pgContext)
+{
+    // learn.microsoft.com/en-us/windows/win32/secauthn/protecting-the-automatic-logon-password
+
+    LSA_HANDLE hPolicy;
+    LSA_UNICODE_STRING Name, *pPwd = NULL;
+    LSA_OBJECT_ATTRIBUTES ObjectAttributes = { sizeof(ObjectAttributes) };
+
+    NTSTATUS status = LsaOpenPolicy(NULL, &ObjectAttributes, 
+                                    POLICY_GET_PRIVATE_INFORMATION, &hPolicy);
+    if (status != STATUS_SUCCESS)
+        return FALSE;
+
+    RtlInitUnicodeString(&Name, L"DefaultPassword");
+    status = LsaRetrievePrivateData(hPolicy, &Name, &pPwd);
+    LsaClose(hPolicy);
+
+    if (status == STATUS_SUCCESS)
+    {
+        if (!SafeGetUnicodeString(pPwd, pgContext->Password,
+                                  sizeof(pgContext->Password) / sizeof(WCHAR)))
+        {
+            status = STATUS_BUFFER_TOO_SMALL;
+        }
+        SecureZeroMemory(pPwd->Buffer, pPwd->Length);
+        LsaFreeMemory(pPwd);
+    }
+    return status == STATUS_SUCCESS;
```

Comment:
Minor suggestion for code consistency (capitalized Status and use NT_SUCCESS macro) :
```suggestion
    NTSTATUS Status = LsaOpenPolicy(NULL, &ObjectAttributes, 
                                    POLICY_GET_PRIVATE_INFORMATION, &hPolicy);
    if (!NT_SUCCESS(Status))
        return FALSE;

    RtlInitUnicodeString(&Name, L"DefaultPassword");
    Status = LsaRetrievePrivateData(hPolicy, &Name, &pPwd);
    LsaClose(hPolicy);

    if (NT_SUCCESS(Status))
    {
        if (!SafeGetUnicodeString(pPwd, pgContext->Password,
                                  sizeof(pgContext->Password) / sizeof(WCHAR)))
        {
            Status = STATUS_BUFFER_TOO_SMALL;
        }
        SecureZeroMemory(pPwd->Buffer, pPwd->Length);
        LsaFreeMemory(pPwd);
    }

    return NT_SUCCESS(Status);
```

Reply:
I don't know if I want to use NT_SUCCESS on the secret. Can it partially read a secret? I need full success or nothing at all.

Reply:
Alright, here's updated suggestion:

Minor suggestion for code consistency (capitalized Status and use NT_SUCCESS macro) :
```suggestion
    NTSTATUS Status = LsaOpenPolicy(NULL, &ObjectAttributes, 
                                    POLICY_GET_PRIVATE_INFORMATION, &hPolicy);
    if (Status != STATUS_SUCCESS)
        return FALSE;

    RtlInitUnicodeString(&Name, L"DefaultPassword");
    Status = LsaRetrievePrivateData(hPolicy, &Name, &pPwd);
    LsaClose(hPolicy);

    if (Status == STATUS_SUCCESS)
    {
        if (!SafeGetUnicodeString(pPwd, pgContext->Password,
                                  sizeof(pgContext->Password) / sizeof(WCHAR)))
        {
            Status = STATUS_BUFFER_TOO_SMALL;
        }
        SecureZeroMemory(pPwd->Buffer, pPwd->Length);
        LsaFreeMemory(pPwd);
    }

    return Status == STATUS_SUCCESS;
```